File: http/http-client-annotation-processor/src/test/java/ru/tinkoff/kora/http/client/annotation/processor/AbstractHttpClientTest.java
package ru.tinkoff.kora.http.client.annotation.processor;

import org.intellij.lang.annotations.Language;
import org.junit.jupiter.api.BeforeEach;
import org.mockito.Mockito;
import ru.tinkoff.kora.annotation.processor.common.AbstractAnnotationProcessorTest;
import ru.tinkoff.kora.config.annotation.processor.processor.ConfigParserAnnotationProcessor;
import ru.tinkoff.kora.config.common.extractor.BooleanConfigValueExtractor;
import ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor;
import ru.tinkoff.kora.config.common.extractor.DoubleArrayConfigValueExtractor;
import ru.tinkoff.kora.config.common.extractor.DurationConfigValueExtractor;
import ru.tinkoff.kora.config.common.extractor.SetConfigValueExtractor;
import ru.tinkoff.kora.config.common.extractor.StringConfigValueExtractor;
import ru.tinkoff.kora.config.common.factory.MapConfigFactory;
import ru.tinkoff.kora.http.client.common.HttpClient;
import ru.tinkoff.kora.http.client.common.declarative.$HttpClientOperationConfig_ConfigValueExtractor;
import ru.tinkoff.kora.http.client.common.request.HttpClientRequest;
import ru.tinkoff.kora.http.client.common.telemetry.*;
import ru.tinkoff.kora.telemetry.common.*;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Flow;
import java.util.function.Function;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public abstract class AbstractHttpClientTest extends AbstractAnnotationProcessorTest {
    protected HttpClient httpClient = mock(HttpClient.class);
    protected HttpClientTelemetry telemetry = mock(HttpClientTelemetry.class);
    protected HttpClientTelemetryFactory telemetryFactory = mock(HttpClientTelemetryFactory.class);
    protected TestObject client;

    @BeforeEach
    public void init() {
        Mockito.reset(httpClient, telemetry, telemetryFactory);
    }

    protected void onRequest(String method, String path, Function<TestHttpClientResponse, TestHttpClientResponse> responseConsumer) {
        when(httpClient.execute(Mockito.argThat(argument -> argument.method().equalsIgnoreCase(method) && argument.uri().toString().equalsIgnoreCase(path))))
            .thenAnswer(invocation -> {
                var f = new CompletableFuture<Void>();
                invocation.getArgument(0, HttpClientRequest.class).body().subscribe(new Flow.Subscriber<ByteBuffer>() {
                    @Override
                    public void onSubscribe(Flow.Subscription subscription) {
                        subscription.request(Long.MAX_VALUE);
                    }

                    @Override
                    public void onNext(ByteBuffer item) {

                    }

                    @Override
                    public void onError(Throwable throwable) {
                        f.completeExceptionally(throwable);
                    }

                    @Override
                    public void onComplete() {
                        f.complete(null);
                    }
                });
                f.get();
                return CompletableFuture.completedFuture(responseConsumer.apply(TestHttpClientResponse.response(200)));
            });
    }

    @Override
    protected String commonImports() {
        return super.commonImports() + """
            import ru.tinkoff.kora.http.client.common.annotation.*;
            import ru.tinkoff.kora.http.client.common.request.*;
            import ru.tinkoff.kora.http.client.common.response.*;
            import ru.tinkoff.kora.http.client.common.*;
            import ru.tinkoff.kora.http.common.annotation.*;
            import ru.tinkoff.kora.http.common.header.*;
            import ru.tinkoff.kora.http.client.common.annotation.HttpClient;
            import ru.tinkoff.kora.http.client.common.interceptor.HttpClientInterceptor;
            import ru.tinkoff.kora.http.client.common.interceptor.HttpClientInterceptor.InterceptChain;
            import reactor.core.publisher.Mono;
            import reactor.core.publisher.Flux;
            import java.util.concurrent.CompletionStage;
            import java.util.concurrent.CompletableFuture;
            """;
    }

    protected TestObject compileClient(List<Object> arguments, @Language("java") String... sources) {
        compile(List.of(new HttpClientAnnotationProcessor(), new ConfigParserAnnotationProcessor()), sources);
        if (compileResult.isFailed()) {
            throw compileResult.compilationException();
        }

        var clientClass = compileResult.loadClass("$TestClient_ClientImpl");
        var durationCVE = new DurationConfigValueExtractor();
//        var telemetryCVE = new $TelemetryConfig_ConfigValueExtractor(
//            new $TelemetryConfig_LogConfig_ConfigValueExtractor(new BooleanConfigValueExtractor()),
//            new $TelemetryConfig_TracingConfig_ConfigValueExtractor(new BooleanConfigValueExtractor()),
//            new $TelemetryConfig_MetricsConfig_ConfigValueExtractor(new BooleanConfigValueExtractor(), new DoubleArrayConfigValueExtractor(c -> c.asNumber().doubleValue()))
//        );
        var telemetryCVE = (ConfigValueExtractor) new $HttpClientTelemetryConfig_ConfigValueExtractor(
            new $HttpClientLoggerConfig_ConfigValueExtractor(new SetConfigValueExtractor<>(new StringConfigValueExtractor()), new BooleanConfigValueExtractor()),
            new $TelemetryConfig_TracingConfig_ConfigValueExtractor(new BooleanConfigValueExtractor()),
            new $TelemetryConfig_MetricsConfig_ConfigValueExtractor(new BooleanConfigValueExtractor(), new DoubleArrayConfigValueExtractor(c -> c.asNumber().doubleValue()))
        );
        var configCVE = new $HttpClientOperationConfig_ConfigValueExtractor(durationCVE, telemetryCVE);

        var configValueExtractor = (ConfigValueExtractor<?>) newObject("$$TestClient_Config_ConfigValueExtractor", telemetryCVE, durationCVE, configCVE);
        var config = configValueExtractor.extract(MapConfigFactory.fromMap(Map.of(
            "url", "http://test-url:8080"
        )).root());


        var realArgs = new ArrayList<>(arguments);
        realArgs.add(0, httpClient);
        realArgs.add(1, config);
        realArgs.add(2, telemetryFactory);
        return this.client = new TestObject(clientClass, realArgs);
    }
}

File: http/http-client-symbol-processor/src/test/kotlin/ru/tinkoff/kora/http/client/symbol/processor/AbstractHttpClientTest.kt
package ru.tinkoff.kora.http.client.symbol.processor

import org.intellij.lang.annotations.Language
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import ru.tinkoff.kora.config.common.extractor.*
import ru.tinkoff.kora.config.common.factory.MapConfigFactory
import ru.tinkoff.kora.config.ksp.processor.ConfigParserSymbolProcessorProvider
import ru.tinkoff.kora.config.ksp.processor.ConfigSourceSymbolProcessorProvider
import ru.tinkoff.kora.http.client.common.HttpClient
import ru.tinkoff.kora.http.client.common.declarative.*
import ru.tinkoff.kora.http.client.common.request.HttpClientRequest
import ru.tinkoff.kora.http.client.common.response.HttpClientResponse
import ru.tinkoff.kora.http.client.common.telemetry.`$HttpClientLoggerConfig_ConfigValueExtractor`
import ru.tinkoff.kora.http.client.common.telemetry.`$HttpClientTelemetryConfig_ConfigValueExtractor`
import ru.tinkoff.kora.http.client.common.telemetry.HttpClientTelemetryFactory
import ru.tinkoff.kora.http.common.body.HttpBody
import ru.tinkoff.kora.ksp.common.AbstractSymbolProcessorTest
import ru.tinkoff.kora.telemetry.common.`$TelemetryConfig_MetricsConfig_ConfigValueExtractor`
import ru.tinkoff.kora.telemetry.common.`$TelemetryConfig_TracingConfig_ConfigValueExtractor`


abstract class AbstractHttpClientTest : AbstractSymbolProcessorTest() {
    val httpResponse = mock<HttpClientResponse>().also {
        whenever(it.code()).thenReturn(200)
        whenever(it.body()).thenReturn(HttpBody.octetStream(ByteArray(0)))
    }
    val httpClient = mock<HttpClient>().also {
        whenever(it.execute(any())).thenReturn(httpResponse)
    }
    val telemetryFactory = mock<HttpClientTelemetryFactory>()
    lateinit var client: TestObject

    protected fun onRequest(method: String, path: String, responseConsumer: (TestHttpClientResponse) -> TestHttpClientResponse) {
        whenever(httpClient.execute(Mockito.argThat { arg: HttpClientRequest -> arg.method().equals(method, ignoreCase = true) && arg.uri().toString() == path }))
            .thenAnswer { invocation ->
                invocation.getArgument(0, HttpClientRequest::class.java).body().close()
                responseConsumer(TestHttpClientResponse(code = 200))
            }
    }


    override fun commonImports(): String {
        return super.commonImports() + """
            import ru.tinkoff.kora.http.client.common.annotation.*
            import ru.tinkoff.kora.http.client.common.request.*
            import ru.tinkoff.kora.http.client.common.response.*
            import ru.tinkoff.kora.http.client.common.*
            import ru.tinkoff.kora.http.common.annotation.*
            import ru.tinkoff.kora.http.client.common.annotation.HttpClient
            import ru.tinkoff.kora.http.common.header.HttpHeaders
            import java.util.concurrent.CompletionStage
            import java.util.concurrent.CompletableFuture
            import ru.tinkoff.kora.http.client.common.interceptor.HttpClientInterceptor
            import ru.tinkoff.kora.http.client.common.interceptor.HttpClientInterceptor.InterceptChain

            """.trimIndent()
    }

    protected fun compile(arguments: List<Any?>, @Language("kotlin") vararg sources: String): TestObject {
        compile0(listOf(HttpClientSymbolProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), *sources)
            .assertSuccess()

        val clientClass = loadClass("\$TestClient_ClientImpl")
        val durationCVE = DurationConfigValueExtractor()
        val telemetryCVE = `$HttpClientTelemetryConfig_ConfigValueExtractor`(
            `$HttpClientLoggerConfig_ConfigValueExtractor`(SetConfigValueExtractor(StringConfigValueExtractor())),
            `$TelemetryConfig_TracingConfig_ConfigValueExtractor`(MapConfigValueExtractor(StringConfigValueExtractor())),
            `$TelemetryConfig_MetricsConfig_ConfigValueExtractor`(DurationArrayConfigValueExtractor(DurationConfigValueExtractor()), MapConfigValueExtractor(StringConfigValueExtractor()))
        )
        val operationTelemetryCVE = `$HttpClientOperationConfig_OperationTelemetryConfig_ConfigValueExtractor`(
            `$HttpClientOperationConfig_OperationTelemetryConfig_LoggingConfig_ConfigValueExtractor`(SetConfigValueExtractor(StringConfigValueExtractor())),
            `$HttpClientOperationConfig_OperationTelemetryConfig_TracingConfig_ConfigValueExtractor`(MapConfigValueExtractor(StringConfigValueExtractor())),
            `$HttpClientOperationConfig_OperationTelemetryConfig_MetricsConfig_ConfigValueExtractor`(
                DurationArrayConfigValueExtractor(DurationConfigValueExtractor()),
                MapConfigValueExtractor<String?>(StringConfigValueExtractor())
            )
        )
        val operationConfigCVE = `$HttpClientOperationConfig_ConfigValueExtractor`(durationCVE, operationTelemetryCVE)

        val configValueExtractor = new("\$\$TestClient_Config_ConfigValueExtractor", telemetryCVE, operationConfigCVE, durationCVE) as ConfigValueExtractor<*>
        val config = configValueExtractor.extract(
            MapConfigFactory.fromMap(
                mapOf(
                    "url" to "http://test-url:8080"
                )
            ).root()
        )


        val realArgs = arrayOfNulls<Any>(arguments.size + 3)
        realArgs[0] = httpClient
        realArgs[1] = config
        realArgs[2] = telemetryFactory
        System.arraycopy(arguments.toTypedArray(), 0, realArgs, 3, arguments.size)
        for ((i, value) in realArgs.withIndex()) {
            if (value is GeneratedObject<*>) {
                realArgs[i] = value.invoke()
            }
        }
        val instance = clientClass.constructors[0].newInstance(*realArgs)
        client = TestObject(clientClass.kotlin, instance)
        return client
    }
}
